### ISSUE TITLE
Submission for InstaHide Challenge - Recovery Attack

### ISSUE THREAD

**Issue by carlini** [Hazelsuko07/InstaHide_Challenge#3]
Attached is our solution to the InstaHide challenge without relying on implementation bugs as found in the PRNG. Please find below the submission for our fully general reconstruction attack. We have significantly improved the visual quality from the arXiv paper, details are upcoming in a longer paper, but the key improvements are described here.

### A description of the attack

Our attack consists of two broad steps.

First, we pair together encoded images to figure out which encoded images correspond to the same image from the original dataset. Then, we use this recovered pairing to reconstruct the original image.

The first step has many sub-steps. We train a neural network to predict if two images likely share an original image, and then evaluate all n^2 images to find the images that are close together and likely share original images. From this we find the cliques of the graph (corresponding to the original images) and form a new bipartite graph mapping each image to its nearest clique. Finally we pair them off.

From here we need to recover the images. We do this by solving the nonlinear system of equations with gradient descent. This is in principle solvable with a MILP but it takes quite a long time, so instead we just go with a heuristic algorithm that seems to work well enough.

Full details are available in [our paper](https://arxiv.org/abs/2011.05315). The attack here has one important difference on top of what is already described in this paper. The reconstruction in this paper results in images with a distinct yellow tint. We found that manually adjusting the color curves could repair this artifact, and so added one additional step to do this automatically.

In particular, we now train a <100 parameter contrast-adjustment neural network that can re-color an image to more closely match the desired color profile.

### The analysis of time complexity

This attack requires O(1) precomputation work to train the neural network similarity detectors.

The remaining attack performs something like O(E^3) if there are E encoded images, but analyzing it completely is somewhat difficult because the constants that matter are orders of magnitude apart.

### The real-world running time (with a specification of the computation platform being used)

This attack takes 10 hours to train the similarity detectors on a P100 GPU. From here the attack takes 2 GPU hours on a P100 and 2 CPU hours. The rough cost of this on a cloud platform is something like $10 to precompute the neural network and $3 to run the remaining attack.

### The 5,000 recovered images as a 5000x32x32x3 numpy array (named as decryption.npy), where each row represents the recovered image of its corresponding encryption in encryption.npy.

Below is a PNG of the 100 original images that constructed the encoded dataset.

![reconstructed](https://user-images.githubusercontent.com/1269300/101229475-5ae4db80-3655-11eb-803c-d1a862e82e1c.png)

I have made available the 5000x32x32x3 numpy file in the format requested (temporarily) available here. 

This first file contains the result of the attack that looks visually best, including the recolorization neural network

https://nicholas.carlini.com/tmp/decryption_realattack.npy

However it's possible that the colorization network makes this visually look better but makes the L2/SSIM worse. I'm not sure what to expect. I would also be curious what how this submission would do:

https://nicholas.carlini.com/tmp/decryption_realattack_raw.npy

### The source code for the attack

Made available here: https://github.com/carlini/privacy/commit/28b8a80924cf3766ab3230b5976388139ddef295

**Comment by Hazelsuko07** [comment]
Hi Nicholas,

Thanks for submitting your attack. We have updated the leaderboard accordingly. It would be nice if you can also attach the training scripts for your similarity detector.

We are looking forward to the details of the last processing step in your upcoming longer paper.

Thanks again!
